### DISCUSSION TITLE
Questions about UniDec processing details

### DISCUSSION THREAD

**Discussion by t-taichi** [michaelmarty/UniDec#66] [Q&A]
Hello. I am a student and would like to understand the details of UniDec's processing.
In UniDec's basic processing, after copying the probabilities of m/z to all charges on the charge axis, I think there is a phase of "allocating peaks to the specific charge".
I couldn't find what this phase was doing in the paper, and I didn't know where it was in the code.
Could anyone please tell me more about it?

**Comment by michaelmarty** [comment]
Hey @t-taichi! 
First, have you checked out the videos and other resources on the wiki page? https://github.com/michaelmarty/UniDec/wiki

Second, I'll do my best to answer, but I may not be understanding exactly what part you are asking about. First, it creates a matrix of m/z vs charge. Each m/z data point has a charge vector. The goal is to distribute the intensity from the m/z data across the available charge states. Ideally, we will have one major charge state with all the intensity. Alternatively, we can have multiple charge states when peaks overlap. The way UniDec does this is by first assuming all charge states in the vector are equally likely. Then, it looks for neighboring charge states at different m/z values. If it sees intensity there, it will raise the intensity for that (m/z, z) point in the matrix. After one iteration of that, it creates a simulated spectrum, compares it with the original data, and then scales each vector to make the simulated data match the actual data. Then, it iterates again with adjusting each (m/z, z) point based on the neighboring (m/z+1, z+1) and (m/z-1, z-1) points, creating another simulated spectrum, and rescaling the whole vector. It reaches convergence when it hits the max iteration limit or stops changing between iterations within some cutoff (I think 0.0001%). 

Let me know what other questions you have. Thanks!
MTM

**Reply by t-taichi** [reply]
Hi @michaelmarty,
Thank you for your so kind answer.
I have seen Wiki and Video. My corrected question is below.

After copying m/z spectrum to all charge, I think UniDec convolve a Boltzmann distribution of valences that diagonally crosses the m/z axis and the charge axis, which gives us a list of m/z vs charges with no width.
Is this work done in "decon_core" of python or "dd_deconv" in c?
I would like to understand the details of this process.

**Reply by michaelmarty** [reply]
I'm not sure it's a Boltzmann distribution, but it's all happening in the deconvolve_interation_speedy in the c code. That should be a good place to start digging.

**Reply by t-taichi** [reply]
Thank you for your information.
The processing I intended was probably within "softargmax".
(Because the delta functions are already obtained when entering deconvolve_iteration_speedy.)
I may ask you again, but I appreciate your continued support.

![image](https://user-images.githubusercontent.com/38335869/178465510-ee8e17b5-8773-444d-8b25-e9f7f8f12c1d.png)
